homeassistant/components/lock/__init__.py

```diff
@@ -87,8 +87,11 @@ async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
 
 async def _async_lock(entity: LockEntity, service_call: ServiceCall) -> None:
     """Lock the lock."""
-    code: str = service_call.data.get(ATTR_CODE, "")
-    if entity.code_format_cmp and not entity.code_format_cmp.match(code):
+    if (
+        (code := service_call.data.get(ATTR_CODE))
+        and entity.code_format_cmp
+        and not entity.code_format_cmp.match(code)
+    ):
```

Comment:
If `code_format` is set, in this case the integration should check if `code` is in the service_call?

Reply:
I believe so and think we should update integrations using ’lock’ once this PR is merged.

Reply:
I agree on this for `_async_lock` and `async_open`, but for `async_unlock` I believe the current code does not need changes. May be it would be an idea to to add attributes like `_attr_unlock_requires_code` , `_attr_lock_requires_code` and `_attr_open_requires_code` (defaulting to false). Then it would be very easy to enable enforce validation.
@emontnemery what do you think?

Reply:
> Your suggestion would not handle the first case which is for example that a `code` is stored in config entry and the user isn't supplying one (example [here](https://github.com/home-assistant/core/blob/dev/homeassistant/components/yale_smart_alarm/lock.py#L53-L55)).

It woulld if we have the attributes for lock , unlock and open.

Reply:
True but I was considering a way to keep the validation as is when user is submitting a code.
However you might be right it's the easiest way to deal with it for the moment and a developer can obviously always implement the validation in it's respective integration.
Let's see if @emontnemery has an alternative

Reply:
Considering all, I think your PR is good. You are right about that integrations need to check for a code if that is needed. May be adding additional attributes does not really add much. I agree merging this change and open a PR to allow MQTT lock to require a code. I still like to hear what @emontnemery has to say about this plan.

Reply:
@frenck, had a chat with @emontnemery what is your opinion about making the code attribute optional in a service call, even when code_format is set? An alternative is add some attributes to be able to set the requirement for a code.

Reply:
Another option would be to add a new entity feature, that demotes the existence of a default code.

For example: `LockEntityFeature.HAS_FALLBACK_DEFAULT_CODE_MAGIC` (obviously needs a better name).

When an integration sets that flag on an lock entity it would:
- Allow omitting the code from the service call
- Validate when a code is passed in the service call

But this would need an architecture discussion IMHO. This can easily derail or extend into: Adding default code as an actual property (which automatically gets handled by HA).

../Frenck

Reply:
https://github.com/home-assistant/core/pull/87588

Reply:
Arch discussion open [here](https://github.com/home-assistant/architecture/discussions/870)
Remaining review comments has now been adjusted

Reply:
@gjohansson-ST your discussion differs from mine (https://github.com/home-assistant/architecture/discussions/872) as you suggest to add a default code to the entity, mine is about the validation of the code, not the code itself.